doc/developer/design/20230411_envelope_upsert_order_by_timestamp.md

```diff
@@ -0,0 +1,108 @@
+- Feature name: Envelope upsert with order by timestamp (limited scope)
+- Associated: https://github.com/MaterializeInc/materialize/issues/16512
+
+# Summary
+
+- Support user specified order by in Kafka upsert sources in a limited capacity. If users want to specify custom ordering which is semantically meaningful to them, there's currently no easy way to do that. They could potentially use `ENVELOPE NONE` along with a `DISTINCT ON` view to get similar results, but it's cumbersome and lacks the automatic compaction on keys with normal `ENVELOPE UPSERT`.
+- Currently, will limit the scope to allow only `ORDER BY` the `TIMESTAMP` metadata field. Depending upon user feedback this can later be expanded to support expressions and extractions from other columns like `HEADERS`.
+
+# Scope
+
+The scope will be limited to only allow `ORDER BY TIMESTAMP` (ascending order).
+
+- In a key-compacted Kafka copic, value = NULL is taken to mean that it’s a delete tombstone and that the corresponding key should be removed. If we want to borrow that tombstone semantic for our more flexible upsert source, it would make sense to limit the `ORDER BY` columns to only the metadata columns.
```

Comment:
```suggestion
- In a key-compacted Kafka topic, value = NULL is taken to mean that it’s a delete tombstone and that the corresponding key should be removed. If we want to borrow that tombstone semantic for our more flexible upsert source, it would make sense to limit the `ORDER BY` columns to only the metadata columns.
```

Reply:
thanks 🙈